pkg/distributor/distributor.go

```diff
@@ -387,20 +392,44 @@ func (d *Distributor) sendSamples(ctx context.Context, ingester *ring.IngesterDe
 	// The use of atomic increments here guarantees only a single sendSamples
 	// goroutine will write to either channel.
 	for i := range sampleTrackers {
-		if err != nil {
-			if atomic.AddInt32(&sampleTrackers[i].failed, 1) <= int32(sampleTrackers[i].maxFailures) {
-				continue
-			}
+		if d.cfg.WaitForAllIngesters {
+			waitForAll(err, sampleTrackers[i], pushTracker)
+		} else {
+			shortCircuit(err, sampleTrackers[i], pushTracker)
+		}
+	}
+}
+
+func shortCircuit(err error, sampleTracker *sampleTracker, pushTracker *pushTracker) {
+	if err != nil {
```

Comment:
Can the error handling be moved up into `sendSamples` so that `waitForAll` and `shortCircuit` use the same error handling?

Reply:
Probably, but there is some subtle logic differences between the two (returning after an error, vs always incrementing sampleTracker.finished). We decided that a little bit of duplication was ok, since it allowed us to leave the previous logic intact. We can refactor it out if you think that is worthwhile however.

Reply:
I think I see what you mean, pushed another commit doing the refactor.